controller/resource.go

```diff
@@ -146,48 +179,209 @@ func (s ServiceStateLister) fetchECSTasks() (map[string]struct{}, error) {
+	}
+
+	if policy != nil {
+		// the policy already exists so return.
+		return nil
+	}
+
+	// create the policy in the local partition and default namespace.
+	_, _, err = s.ConsulClient.ACL().PolicyCreate(&api.ACLPolicy{
+		Name:        xnsPolicyName,
+		Description: xnsPolicyDesc,
+		Partition:   s.Partition,
+		Namespace:   DefaultNamespace,
+		Rules:       fmt.Sprintf(xnsPolicyTpl, s.Partition),
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("creating cross-namespace policy: %w", err)
+	}
+
+	s.Log.Info("created cross-namespace policy", xnsPolicyName)
+	return nil
+}
+
+// createNamespaces creates each namespace in the list.
+// It also attaches the cross-namespace read policy to the namespace so that
+// each token created in the namespace inherits that policy.
+//
+// It does nothing for namespaces that already exist.
+func (s ServiceStateLister) createNamespaces(ns []string) error {
+	// retrieve the list of existing namespaces
+	existingNS, _, err := s.ConsulClient.Namespaces().List(&api.QueryOptions{Partition: s.Partition})
+	if err != nil {
+		return err
+	}
+
+	// create any namespaces that do not already exist
+	create := make([]string, 0, len(ns))
+	for _, n := range ns {
+		for _, e := range existingNS {
+			if e.Name != n {
+				create = append(create, n)
+				break
+			}
+		}
+	}
+	failed := make([]string, 0, len(create))
+	for _, n := range create {
+		s.Log.Info("creating namespace", n)
+		_, _, err = s.ConsulClient.Namespaces().Create(&api.Namespace{
+			Partition: s.Partition,
+			Name:      n,
+			ACLs:      &api.NamespaceACLConfig{PolicyDefaults: []api.ACLLink{{Name: xnsPolicyName}}},
+		}, nil)
+		if err != nil {
+			s.Log.Error("creating namespace", n, err)
+			failed = append(failed, n)
+		}
+	}
+	if len(failed) > 0 {
+		return fmt.Errorf("namespace creation failed for %s", strings.Join(failed, ","))
+	}
+	return nil
 }
 
-func (s ServiceStateLister) newServiceInfo(name string, serviceState ServiceState) *ServiceInfo {
+func (s ServiceStateLister) newServiceInfo(serviceName QName, serviceState ServiceState) *ServiceInfo {
 	return &ServiceInfo{
 		SecretsManagerClient: s.SecretsManagerClient,
 		ConsulClient:         s.ConsulClient,
 		Cluster:              s.Cluster,
 		Log:                  s.Log,
 		SecretPrefix:         s.SecretPrefix,
-		ServiceName:          name,
+		QName:                serviceName,
 		ServiceState:         serviceState,
 	}
 }
 
+// Task definition ARN looks like this: arn:aws:ecs:us-east-1:1234567890:task-definition/service:1
+func (s ServiceStateLister) serviceNameForTask(t *ecs.Task) (QName, error) {
+	var partition, namespace string
+	if PartitionsEnabled(s.Partition) {
+		partition = tagValue(t.Tags, partitionTag)
+		namespace = tagValue(t.Tags, namespaceTag)
+	}
+	if serviceName := tagValue(t.Tags, serviceNameTag); serviceName != "" {
+		return NewQName(partition, namespace, serviceName), nil
+	}
+	taskDefArn := *t.TaskDefinitionArn
+	splits := strings.Split(taskDefArn, "/")
```

Comment:
nit: fmt.Sscanf could also make this more readable (but I realize you didn't write this portion)

Reply:
I'm going to leave this as is for now... While I agree that it could make the code more readable, I played around with `fmt.Scanf` a bit and it seems to struggle with a minimal format specifier like `"%s/%s:%s"`.. another option could be a `regexp` match with positional results.